### DISCUSSION TITLE
sync dependency on async router, how does this work?

### DISCUSSION THREAD

**Discussion by mr-mapache** [fastapi/fastapi#11991] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def users_repository() -> Users:
    raise NotImplementedError("Override this dependency with a concrete implementation")

def user_factory() -> User:
    raise NotImplementedError("Override this dependency with a concrete implementation")

router = APIRouter()

@router.post("/users", response_model=UserResponse)
async def create_user(user: User = Depends(user_factory), users: Users = Depends(users_repository)):
    try:
        await users.add(user)
        return user
    
    except exceptions.UserAlreadyExists:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="User already exists")
    
    except Exception:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Internal server error")
```

### Description

For example in the code, I'm using dependency injection a lot for decoupling my domain layer and my controller layer, like the example code, I have a users repository and a user aggregate factory, that will be injected later in the code, and I have a a questions regarding this.

Is it okay to override sync dependencies with async ones? 

Does fastapi add sync dependencies in executor mode from asyncio or it just plain use them? I don't want to be adding a lot of threads simply for just creating a domain object or something is not a heavy cpu task, but also don't want to lock my code with sync dependencies, for example here: 

```python  

async def hash_credential(form: OAuth2PasswordRequestForm = Depends(), security: Security = Depends(security_service)) -> Credential:
    loop = get_event_loop()
    hashed_password = await loop.run_in_executor(security.hash, form.password)
    return Credential(username=form.username, password=hashed_password)


@router.post("/users/{id}/credentials", response_model=UserResponse)
async def add_credentials(id: UUID, credential: Credential = Depends(hash_credential), users: Users = Depends(users_repository)):
    user = await users.get_by_id(id)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
    await users.credentials.add(credential, user)
    return user
```

I need the passwords to be hashed before it even touch the backend because, one can't really trust other people working on something, hashing the password is a heavy CPU task so it must be added to a run in executor mode from asyncio, but I was having dobouts if I have a sync dependency like that, will it be added into the code in executor mode? or will be just used without asyncio?

### Operating System

Linux

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

Python 3.12.3

**Comment by YuriiMotov** [comment, marked as answer]
> Is it okay to override sync dependencies with async ones?

There is no problem with that. FastAPI will see that this function is `sync` and run it in the thread_pool.

> Does fastapi add sync dependencies in executor mode from asyncio or it just plain use them?

Sync dependencies will be executed in separate thread and will not block the event loop.


If your dependency function is not a heavy CPU task, just use `async` to define it

**Reply by mr-mapache** [reply]
Thanks! I was doing the oposite, and making my more code complicated because of that detail.
